<a href="https://colab.research.google.com/github/Vynogradova/homework/blob/main/Homework_9_5_%22apply%2C_groupby%2C_pivot_table_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
import pandas as pd

In [6]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/supermarket_sales.csv')

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [7]:
branch_profit = df.groupby('Branch')['gross income'].sum().sort_values(ascending=False)

print(branch_profit)

most_profitable_branch = branch_profit.idxmax()
print(f"\nНайприбутковіший філіал: {most_profitable_branch}")

Branch
C    5265.1765
A    5057.1605
B    5057.0320
Name: gross income, dtype: float64

Найприбутковіший філіал: C


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [9]:
branch_profit = df.groupby('Branch')['gross income'].sum()

top_branch = branch_profit.idxmax()

branch_city = df[df['Branch'] == top_branch]['City'].unique()[0]

print(f"Найприбутковіший філіал: {top_branch}")
print(f"Місто, де він знаходиться: {branch_city}")

Найприбутковіший філіал: C
Місто, де він знаходиться: Naypyitaw


3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [10]:
invoices_by_category = df.pivot_table(
    index='Branch',
    columns='Product line',
    values='Invoice ID',
    aggfunc='count'
)

# Вивід результату
print(invoices_by_category)

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45  


In [12]:
styled_table = invoices_by_category.style.set_table_styles(
    [{'selector': 'td, th', 'props': [('background-color', 'white'), ('color', 'black')]}]
)
styled_table

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [13]:
top_branch_electronics = invoices_by_category['Electronic accessories'].idxmax()
top_count = invoices_by_category['Electronic accessories'].max()

print(f"Найбільше інвойсів з категорії 'Electronic accessories' у філіалі: {top_branch_electronics}")
print(f"Кількість інвойсів: {top_count}")

Найбільше інвойсів з категорії 'Electronic accessories' у філіалі: A
Кількість інвойсів: 60


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


Гіпотеза 1: Чи є різниця у середньому чеку між чоловіками та жінками?

In [14]:
avg_total_by_gender = df.groupby('Gender')['Total'].mean().round(2)
print(avg_total_by_gender)

Gender
Female    335.10
Male      310.79
Name: Total, dtype: float64


Жінки мають вищий середній чек (335.10), ніж чоловіки (310.79).

**Як використовувати:**

1. Зосередитись на маркетингових кампаніях для жінок, пропонуючи персоналізовані акції та комплекти товарів.

2. Для чоловіків розробити стимули, що підвищують середній чек (наприклад бонуси).

Гіпотеза 2: У який день тижня відбувається найбільше покупок?

In [15]:
df['Date'] = pd.to_datetime(df['Date'])
df['Weekday'] = df['Date'].dt.day_name()

purchases_by_day = df['Weekday'].value_counts()
print(purchases_by_day)

Weekday
Saturday     164
Tuesday      158
Wednesday    143
Friday       139
Thursday     138
Sunday       133
Monday       125
Name: count, dtype: int64


Найбільше покупок відбувається у суботу (164), а найменше — у понеділок (125).

**Як використовувати:**

1. Планувати акції та рекламу на суботу, щоб максимізувати продажі.

2. Проводити спеціальні стимули у понеділок для підвищення активності у цей день.

3. Регулювати графік роботи персоналу, орієнтуючись на пікові дні продажів.

3 Гіпотеза: Які способи оплати переважають у кожному місті.

In [16]:
payment_by_city = df.pivot_table(
    index='City',
    columns='Payment',
    values='Invoice ID',
    aggfunc='count'
)
print(payment_by_city)

Payment    Cash  Credit card  Ewallet
City                                 
Mandalay    110          109      113
Naypyitaw   124           98      106
Yangon      110          104      126


Аналіз способів оплати в різних містах показує:

- У Mandalay способи оплати розподілені майже рівномірно: Cash (110), Credit card (109), Ewallet (113).
- У Naypyitaw найбільше покупок оплачують готівкою (124), менше — Ewallet (106) та кредитною карткою (98).
- У Yangon найбільше використовується Ewallet (126), далі Cash (110) та Credit card (104).

**Як використовувати:**

- Для Yangon варто розвивати цифрові методи оплати, впроваджувати акції з використанням Ewallet.
- У Naypyitaw можна стимулювати безготівкові платежі, розширюючи можливості кредитних карток та мобільних гаманців.
- В Mandalay ефективно підтримувати всі способи оплати, адже попит розподілений рівномірно.